### ISSUE TITLE
SwiftProtobuf plugin for SwiftPM?

### ISSUE THREAD

**Issue by kielgillard** [apple/swift-protobuf#1207]
tl;dr: Is it possible for this repo to offer a canonical SwiftProtobuf plugin for SwiftPM, [as demonstrated here](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md#example-2-swiftprotobuf)?

Now that [Swift 5.6 is out in the Xcode 13.3 betas and later](https://developer.apple.com/documentation/xcode-release-notes/xcode-13_3-release-notes), Swift Packages now support build tool plugins, as defined in [SE-0303](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md) and SE-0325.

However, for the time being, it seems like everyone will have to roll their own protobuf to Swift generator plugin. It doesn't seem like there is a plugin already "out there" and one does not simply copy and paste the example from the proposal. Speaking for myself, I am not that familiar with the tool and I would certainly have no idea how to support a hand rolled plugin going forward.

Is it possible for this repo to offer a canoncial SwiftProtobuf plugin for SwiftPM? All I really want to do is declare a dependency to this repo and add a plug in to my target. Is this straightforward usability naive?

Some details about my use case. I just have some protobuf files organised into folders that I want to generate Swift from. I want to compile that source into a library. This library is a dependency among other targets in different packages.

macOS Monterey 12.2 (21D49)
Xcode 13.3 / Build version 13E5086k
swift-driver version: 1.44.2 Apple Swift version 5.6 (swiftlang-5.6.0.320.8 clang-1316.0.18.8)
Target: x86_64-apple-macosx12.0
https://github.com/apple/swift-protobuf/releases/tag/1.18.0

**Comment by abertelrud** [comment]
There is a prototype in the `eng/add-build-tool-plugin` branch in a fork at https://github.com/abertelrud/swift-protobuf, and a use case in https://github.com/abertelrud/swiftpm-buildtool-plugin-examples that could be used as a start.  It's not robust enough to be real yet but it would be great to work with someone on real support.

**Comment by Lukasa** [comment]
I should note that it will still be required to get hold of `protoc` outside of this repo. However, I'm +1 on upstreaming this support here.

**Comment by abertelrud** [comment]
Yes, that's true — the prototype interns it but it will definitely need be done using a binary target instead.  There was no technical reason to have to put it inside the prototype repo, it was just expedient since I didn't have a good place to host a binary.

**Comment by tbkka** [comment]
We support multiple operating systems and `protoc` is a moving target, so it makes no sense for us to be involved in distributing `protoc` binaries. Unless SwiftPM can check out and build `protoc` from source, we'll have to settle for documenting this prerequisite.

**Comment by thomasvl** [comment]
aside - is this likely a dup of #1115 

Since they publish a protoc binary for each release, I wonder if there's a way to do a _protoc Package_ that somehow wraps using those downloads for their supported platforms?

**Comment by Lukasa** [comment]
I don’t think SwiftPM has the feature set for this yet. But it would be good to enumerate what is missing.

**Comment by thomasvl** [comment]
> I don’t think SwiftPM has the feature set for this yet. But it would be good to enumerate what is missing.

Yea, I guess SE-0303 calls out the sandbox blocking network access. It does mention limiting file system access would could also be an issue for having users provide the binaries.

Thinking a little more about this, any ideas how SE-0303 is supposed to work with Swift Playgrounds on an iPad?  If a plugin just needs to read files and generate code, they I guess it works, but if a plugin needs to run other tools, how is that supposed to work? That's basically the category this project falls into because of the need to parse the `.proto` files.

**Comment by kielgillard** [comment]
So the impression I'm getting is that it's not possible to do because the protobuf compiler needs to be downloaded or built and the kinds of clients the plug-in might have can't download it or SwiftPM can't build it. Is this right? Is there any point offering to help?

If there is any point to it, I'm looking into the possibility of helping contribute something but I -- or maybe anyone else that wants to help, for that matter -- need a bit of guidance.

Is @abertelrud's fork a good place to start? Can someone please list the things that need to be done to make it acceptable (since it was claimed it was not yet)? I can work with a list but I've no idea on my own.

**Comment by tbkka** [comment]
It's partly possible, probably worth doing, @abertelrud's fork should be a good place to start, and we really do appreciate the offer to help.  ;-)

The result won't be perfect, since SwiftPM won't be able to automatically install and/or build the C++ `protoc` binary that's required.  So the challenge here is probably to figure out how clear you can make the failure case.

**Comment by thomasvl** [comment]
I'm somewhat swamped with work things at the moment, could one of the interested folk revive or start a new forums thread about this sorta need for plugins to help raise the issue to see if other things might have similar issues and get folks thinking about solutions?

**Comment by d-date** [comment]
The [SE-0305](https://github.com/apple/swift-evolution/blob/main/proposals/0305-swiftpm-binary-target-improvements.md) supports Artifact archive bundle format for non-Swift CLI tools like `protoc`. It has already merged into Swift 5.6.

**Comment by kielgillard** [comment]
@abertelrud Is @d-date's observation really something a protobuf plug in could use? How come you didn't suggest it before?

**Comment by thomasvl** [comment]
> The [SE-0305](https://github.com/apple/swift-evolution/blob/main/proposals/0305-swiftpm-binary-target-improvements.md) supports Artifact archive bundle format for non-Swift CLI tools like `protoc`. It has already merged into Swift 5.6.

Sounds like it might work for supporting macOS & Linux based protoc binaries, but the plugin be able to support Swift Playground usage on an iPad.  Unless someone else knows a way to make a protoc for that case.

**Comment by FranzBusch** [comment]
I haven't looked at Playgrounds yet but will definitely give them a run after coming back from my vacation.

I also do not really like providing a path. So far Xcode tolerates that since we pass the path to SPM to actually look up the tool at the path.

IMO, the best approach would be if we could get an artefactbundle hosted over in the protobuf repo. @thomasvl Any idea if that would be possible?

**Comment by tbkka** [comment]
I'm fine with a path to the protoc binary.  I don't know that we can do substantially better than this with the current tools, so I'm okay with something that doesn't work for Xcode Playgrounds on iPad as long as it does work for desktop Xcode as well as command-line users on Linux and Windows.

I'm much more interested in the target dependencies issue Franz mentioned above.  What exactly is the issue?

**Comment by FranzBusch** [comment]
I also think that not supporting playgrounds initially is okay. We need to see if we can fix that in SPM. IMO artifactbundles are the proposed way by SPM folks so we should go ahead with them.

@tbkka The target dependency issue is the following:
- Target A: Generates public symbols from proto file `Foo.proto`
- Target B: Generates symbols from proto file `Bar.proto` which depend on messages declared in `Foo.proto`

Two interesting things to tackle here:
1. How does a valid import statement look like?
2. How do we generate the correct ProtoPathModuleMappings?

## Import statements
With my current state of the plugin you can just write `import "Foo.proto";` in target B and it properly generates the symbols. This is due to the fact that I include every directory of a recursive transitive dependency in the proto file lookup

```swift
  // We include all directories of all transitive dependencies as proto file search paths.
  var protoSearchPaths = [Path]()
  protoSearchPaths.append(target.directory)
  for dependency in target.recursiveTargetDependencies {
      protoSearchPaths.append(dependency.directory)
  }
  protocArgs.append(contentsOf: protoSearchPaths.flatMap { ["-I", "\($0)"] })
```

This works, but I don't think this is the best solution since it leads to very easy conflicts in larger dependency trees. One potential solution here might be to instead use the parent folder of each dependencies target directory. This would result in imports like this `import targetA/Foo.proto;`. However, the downside here is that any package that has non-standard file layouts breaks. Furthermore, moving files inside a package also becomes a breaking change. I am open to suggestions here!


## Generate ProtoPathModuleMappings
This one is a bit tricky since every target can have multiple configurations for protoc. For proto file that is included in the generation of a public symbol needs to be included in a generated `ProtoPathModuleMappings` file. Furthermore, every plugin invocation would need to merge all mapping files of its recursive transitive dependencies. This is the second thing that is necessary to allow dependencies between targets that generate symbols.


I think we first need to tackle 1. and discuss how we want to support this. Solving the mapping file generation is relatively straight forward.

**Comment by FranzBusch** [comment]
@thomasvl @tbkka What do you think about punting support for "multi target proto generation with dependencies" from the first release. I think it would be already super valuable to have a SPM plugin that works for single target generation and we can always extend that later on.

If you both agree we can go forward with that, then I would change the proposal a bit and try to get everything ready for a PR.
